mantidproject/mantid
# Lagrange: user feedback features

**Description of work.**

This PR implements issues and requests reported by the scientific user, namely:
- [x] Bug: Water correction does not take into account the difference between energy exchange or initial energy situations leading to interpolation going to the region with no data has been addressed consistently with the sample data.
- [x] More workspaces are now be available for debugging, notably the water correction, both pre- and post-interpolation
- [x] User now has the ability to choose normalisation from: None, and Monitor. Time will be added after the switch to NXS.
- [x] Metadata of time and temperature is now loaded from the ASCII and added to the sample logs.

The support for loading NXS files will be added as soon as the format is decided. 

**To test:**
1. Code review.
2. All tests should be passing
3. Run the example added to the documentation (make sure you have LAGRANGE unit test data in your search directories):
```
# full correction of a single monochromator scan, with multiple files
result = LagrangeILLReduction(SampleRuns='012869:012871',
                                  ContainerRuns='012882:012884',
                                  CorrectionFile='correction-water-cu220-2020.txt',
                                  UseIncidentEnergy=False,
                                  ConvertToWaveNumber=False)
 run = result.getRun()
 times = run.getLogData('time').value
 eis = result.readX(0)
 temperatures = run.getLogData('temperature').value
 plt.plot(eis, temperatures)
 plt.xlabel("Ei (meV)")
 plt.ylabel("temperature (K)")
 plt.show()
```
A proper figure should appear with temperature as a function of initial energy displayed correctly. There should be a rapid change at the lowest energies and plateau towards high-Ei.

Fixes #34837. 






---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
